File: LATG/DB/Schema.hs
-- oh god why
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE NamedFieldPuns #-}

module LATG.DB.Schema where

import Data.Profunctor.Product (p10, p13, p15)
import Opaleye
  ( Field,
    FieldNullable,
    SqlArray,
    SqlInt2,
    SqlInt4,
    SqlText,
    SqlTimestamptz,
    SqlUuid,
    Table,
    optionalTableField,
    readOnlyTableField,
    requiredTableField,
    table,
  )

documents ::
  Table
    ( (),
      Field SqlUuid,
      Field SqlTimestamptz,
      Field SqlTimestamptz,
      Maybe (FieldNullable SqlTimestamptz),
      Field SqlText,
      Field SqlText,
      Maybe (FieldNullable SqlText),
      Maybe (FieldNullable SqlText),
      Maybe (FieldNullable SqlText)
    )
    ( Field SqlInt4,
      Field SqlUuid,
      Field SqlTimestamptz,
      Field SqlTimestamptz,
      FieldNullable SqlTimestamptz,
      Field SqlText,
      Field SqlText,
      FieldNullable SqlText,
      FieldNullable SqlText,
      FieldNullable SqlText
    )
documents =
  table
    "documents"
    ( p10
        ( readOnlyTableField "id",
          requiredTableField "uuid",
          requiredTableField "created_date",
          requiredTableField "published_date",
          optionalTableField "updated_date",
          requiredTableField "canonical_url",
          requiredTableField "doc_type",
          optionalTableField "content",
          optionalTableField "colophon",
          optionalTableField "page_src_url"
        )
    )

entries ::
  Table
    ( (),
      Field SqlInt4,
      Field SqlInt4,
      Field SqlInt4,
      Field SqlInt4,
      Field SqlInt4,
      Maybe (FieldNullable SqlText),
      Maybe (FieldNullable SqlText),
      Maybe (FieldNullable SqlText),
      Maybe (FieldNullable SqlText),
      Maybe (Field (SqlArray SqlText)),
      Maybe (Field (SqlArray SqlText)),
      Maybe (FieldNullable SqlText),
      Maybe (FieldNullable SqlText),
      Maybe (FieldNullable SqlText)
    )
    ( Field SqlInt4,
      Field SqlInt4,
      Field SqlInt4,
      Field SqlInt4,
      Field SqlInt4,
      Field SqlInt4,
      FieldNullable SqlText,
      FieldNullable SqlText,
      FieldNullable SqlText,
      FieldNullable SqlText,
      Field (SqlArray SqlText),
      Field (SqlArray SqlText),
      FieldNullable SqlText,
      FieldNullable SqlText,
      FieldNullable SqlText
    )
entries =
  table
    "entries"
    ( p15
        ( readOnlyTableField "id",
          requiredTableField "document",
          requiredTableField "year",
          requiredTableField "month",
          requiredTableField "day",
          requiredTableField "ordinal",
          optionalTableField "short_name",
          optionalTableField "summary",
          optionalTableField "location",
          optionalTableField "short_name",
          optionalTableField "photos",
          optionalTableField "reply_to",
          optionalTableField "repost_of",
          optionalTableField "rsvp",
          optionalTableField "rsvp_to"
        )
    )

projects ::
  Table
    ( (),
      Field SqlInt4,
      Field SqlText,
      Field SqlText,
      Field SqlInt2,
      Field SqlTimestamptz,
      Maybe (FieldNullable SqlTimestamptz),
      Field SqlTimestamptz,
      Field SqlText,
      Maybe (FieldNullable SqlText),
      Maybe (FieldNullable SqlText),
      Maybe (FieldNullable SqlText),
      Maybe (FieldNullable SqlText)
    )
    ( Field SqlInt4,
      Field SqlInt4,
      Field SqlText,
      Field SqlText,
      Field SqlInt2,
      Field SqlTimestamptz,
      FieldNullable SqlTimestamptz,
      Field SqlTimestamptz,
      Field SqlText,
      FieldNullable SqlText,
      FieldNullable SqlText,
      FieldNullable SqlText,
      FieldNullable SqlText
    )
projects =
  table
    "projects"
    ( p13
        ( readOnlyTableField "id",
          requiredTableField "document",
          requiredTableField "slug",
          requiredTableField "status",
          requiredTableField "featured_order",
          requiredTableField "started_date",
          optionalTableField "finished_date",
          requiredTableField "sort_date",
          requiredTableField "name",
          optionalTableField "summary",
          optionalTableField "url",
          optionalTableField "source",
          optionalTableField "location"
        )
    )

File: cms/src/schema.rs
table! {
    content (id) {
        id -> Int4,
        content -> Text,
        content_type -> Text,
        rendered_html -> Nullable<Text>,
    }
}

table! {
    post (id) {
        id -> Int4,
        uuid -> Uuid,
        year -> Int4,
        month -> Int4,
        day -> Int4,
        ordinal -> Int4,
        title -> Nullable<Text>,
        description -> Nullable<Text>,
        slug -> Nullable<Text>,
        created_date -> Timestamp,
        published_date -> Nullable<Timestamp>,
        updated_date -> Nullable<Timestamp>,
        reply_to -> Nullable<Text>,
        bookmark_of -> Nullable<Text>,
        like_of -> Nullable<Text>,
        repost_of -> Nullable<Text>,
        rsvp -> Nullable<Text>,
        location -> Nullable<Text>,
        content -> Int4,
    }
}

table! {
    project (id) {
        id -> Int4,
        uuid -> Nullable<Uuid>,
        title -> Text,
        description -> Text,
        slug -> Text,
        start_date -> Int4,
        end_date -> Nullable<Int4>,
        published_date -> Nullable<Int4>,
        updated_date -> Nullable<Int4>,
        url -> Nullable<Text>,
        status -> Nullable<Text>,
        source_urls -> Nullable<Text>,
        thumbnail -> Nullable<Text>,
        content -> Int4,
    }
}

table! {
    tag (slug) {
        slug -> Text,
        name -> Nullable<Text>,
        background_color -> Nullable<Text>,
        color -> Nullable<Text>,
        description -> Text,
    }
}

table! {
    tag_to_object (object_uuid, tag) {
        object_uuid -> Uuid,
        tag -> Text,
    }
}

table! {
    tagged_object (uuid) {
        uuid -> Uuid,
        url -> Nullable<Text>,
    }
}

joinable!(post -> content (content));
joinable!(post -> tagged_object (uuid));
joinable!(project -> content (content));
joinable!(project -> tagged_object (uuid));
joinable!(tag_to_object -> tag (tag));
joinable!(tag_to_object -> tagged_object (object_uuid));

allow_tables_to_appear_in_same_query!(
    content,
    post,
    project,
    tag,
    tag_to_object,
    tagged_object,
);
